src/Functions/FunctionsConversion.cpp

```diff
@@ -3736,20 +3737,41 @@ class FunctionCast final : public IFunctionBase
             element_wrappers.reserve(to_names.size());
             to_reverse_index.reserve(from_names.size());
 
+            size_t num_from_fields = 0;
             for (size_t i = 0; i < to_names.size(); ++i)
             {
                 auto it = from_positions.find(to_names[i]);
                 if (it != from_positions.end())
                 {
                     element_wrappers.emplace_back(prepareUnpackDictionaries(from_element_types[it->second], to_element_types[i]));
                     to_reverse_index.emplace_back(it->second);
+                    ++num_from_fields;
                 }
                 else
                 {
                     element_wrappers.emplace_back();
                     to_reverse_index.emplace_back();
                 }
             }
+
+            bool strict_named_tuple_conversion = false;
+
+            if (DB::CurrentThread::isInitialized())
+            {
+                const DB::ContextPtr query_context = DB::CurrentThread::get().getQueryContext();
+
+                // Avoid strict checks for background operations like part
```

Comment:
What operations are you trying to cover here? I'm not sure what constitutes background operations but this might hide things in Refreshable Materialized Views which I don't know if it's desirable.

Reply:
It's added to fix test `01825_type_json_3.sql.j2`, which somehow does `Tuple(_dummy UInt8) -> Tuple(...)` during merge. It will also help make related mutations work. Anyway I cannot find a reasonable way of interacting this setting with background jobs.

Reply:
What do you think about adding 3 options just in case:
- 0: Always fills missing values with default values.
- 1: Only fills missing values in background operations. It throws in foreground operations
- 2: Always throws

Then we leave the default to 1, and `{"strict_named_tuple_conversion", 0, 1, "New setting"}` in history, so old compatibility works as before and new ones only throws in foreground. In the future if we improve background operations we can switch to 2 (I'm expecting the issues might come from not considering it important because we didn't use named tuples that much before).

Reply:
Yeah, that sounds solid.

Reply:
BTW, just checked the failures in private and it's failing because of this. You can't guarantee that the merge will happen in the background process, so it will fail when calling optimize: `Some fields from source tuple are lost when casting Tuple(_dummy UInt8) to Tuple(k1 Nested(k2 String, k3 String)) (strict_named_tuple_conversion is enabled)`

Seems related to the object type, but I don't know the internals to propose a good solution.

@Avogar How are we handling types and conversion inside the old Object type? It seems we are using some kind of tuple and then we cast it to something else during merges:

https://github.com/ClickHouse/ClickHouse/blob/67c1e89d90ef576e62f8b1c68269742a3c6f9b1e/src/DataTypes/ObjectUtils.cpp#L497-L499